### PR TITLE
refactor clients

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] The PR targets the `latest` branch.
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] Apache v2 license headers are included in each file.
 
[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/master/CHANGELOG.md

### CODE REVIEW

File: sdk/storage/azblob/service.go [hunk 1]
```diff
@@ -15,32 +15,78 @@ const (
...
+const (
+	// V2019_02_02 specifies the 2019-02-02 service API version.
+	V2019_02_02 = "2019-02-02"
+)
+
+//ServiceClient provides access to Azure Storage service resources and blob containers.
 type ServiceClient struct {
 	u *url.URL
 	p azcore.Pipeline
+	o ServiceClientOptions
```
Comment (JeffreyRichter) [inline sdk/storage/azblob/service.go]: I assume this was mostly added due to ServiceVersion? Maybe we should just put that here and let the rest of the options be GC'd.

File: sdk/storage/azblob/service.go [hunk 2]
```diff
@@ -15,32 +15,78 @@ const (
...
+	Version ServiceVersion
+
+	// SecondaryEndpoint is the secondary storage endpoint that can be read from if the storage
+	// account is enabled for RA-GRS.
+	// If this value is set, the secondary endpoint will be used for GET or HEAD requests during retries.
+	// If the status of the response from the secondary endpoint is a 404, then subsequent retries for
+	// the request will not use the secondary endpoint again, as this indicates that the resource
+	// may not have propagated there yet. Otherwise, subsequent retries will alternate back and forth
+	// between the primary and secondary endpoint.
+	SecondaryEndpoint string
```
Comment (JeffreyRichter) [inline sdk/storage/azblob/service.go]: In Track 1.8, this is a retry policy option and I think that is where this belongs; not here. Why was this moved here?

File: sdk/storage/azblob/service.go [hunk 3]
```diff
@@ -15,32 +15,78 @@ const (
...
+}
+
+func setDefaultClientOptions(s *ServiceClientOptions) *ServiceClientOptions {
+	if s == nil {
+		s = &ServiceClientOptions{}
+	}
+	if len(s.Version) == 0 {
```
Comment (JeffreyRichter) [inline sdk/storage/azblob/service.go]: Can't you compare against "" instead - feels more natural to me.

File: sdk/storage/azblob/service.go [hunk 4]
```diff
@@ -15,32 +15,78 @@ const (
...
+// Pass nil for options to use the default configuration.
+func NewServiceClient(endpoint string, cred azcore.Credential, options *ServiceClientOptions) (*ServiceClient, error) {
+	if cred == nil {
+		cred = azcore.AnonymousCredential()
+	}
+	options = setDefaultClientOptions(options)
+	p := azcore.NewPipeline(options.HTTPTransport,
 		azcore.NewTelemetryPolicy(options.Telemetry),
 		azcore.NewUniqueRequestIDPolicy(),
 		azcore.NewRetryPolicy(options.Retry),
```
Comment (JeffreyRichter) [inline sdk/storage/azblob/service.go]: This should be a RetryPolicy that is unique to azblob; not the azcore RetryPolicy.
The azblob policy takes a SecondaryEndpoint in its options

File: sdk/storage/azblob/service.go [hunk 5]
```diff
@@ -15,32 +15,78 @@ const (
...
 		cred.AuthenticationPolicy(azcore.AuthenticationPolicyOptions{Scopes: []string{scope}}),
 		azcore.NewRequestLogPolicy(options.LogOptions))
-	return NewServiceClientWithPipeline(endpoint, p)
+	return NewServiceClientWithPipeline(endpoint, p, options)
```
Comment (JeffreyRichter) [inline sdk/storage/azblob/service.go]: OK, I think I see the issue. lets create a ServiceApiVersion pipeline policy and add it to the pipeline. It is configured with the service api version. Now, we don't need to add options fields to any XxxClient and we don't pass options to any XxxClientWithPipeline functions. 

The options are for configuring the pipeline. It doesn't feel right to have a configured pipeline AND then to also have to pass options to a XxxClientWithPipeline function.

File: sdk/storage/azblob/service.go [hunk 6]
```diff
@@ -15,32 +15,78 @@ const (
...
-	return &ServiceClient{u: u, p: p}, nil
+	options = setDefaultClientOptions(options)
+	return &ServiceClient{u: u, p: pipeline, o: *options}, nil
 }
 
-func (c *ServiceClient) ServiceVersion() string {
-	// this is a method instead of a package-level const to handle the case of composite services
-	return "2018-11-09"
+// ServiceVersion returns the service API version used when making requests.
+func (c *ServiceClient) ServiceVersion() ServiceVersion {
```
Comment (JeffreyRichter) [inline sdk/storage/azblob/service.go]: I don't think we need this function at all. Do you disagree?
If you think we do, then we'd need a way to return ALL of options and I really don't think this is necessary.

File: sdk/storage/azblob/service.go [hunk 7]
```diff
@@ -95,7 +141,7 @@ func (c *ServiceClient) listContainersPreparer(options *ListContainersOptions) *
 		}
 	}
 	req.SetQueryParam("comp", "list")
-	req.Header.Set("x-ms-version", c.ServiceVersion())
+	req.Header.Set("x-ms-version", string(c.ServiceVersion()))
```
Comment (JeffreyRichter) [inline sdk/storage/azblob/service.go]: This would be in a pipeline policy now.

File: sdk/storage/azblob/service.go (hunk 6 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/service.go]: Agreed it can be removed.

File: sdk/storage/azblob/service.go (hunk 3 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/service.go]: Will fix.

**Comment by jhendrixMSFT** [comment]
Closing as no longer needed.
